federation/planner_test.go

```diff
@@ -0,0 +1,407 @@
+package federation
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/samsarahq/thunder/graphql"
+	"github.com/samsarahq/thunder/graphql/schemabuilder"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func mustParse(s string) *graphql.SelectionSet {
+	return graphql.MustParse(s, map[string]interface{}{}).SelectionSet
+}
+
+func setupExecutor(t *testing.T) (*Executor, error) {
+	schemas := map[string]map[string]*schemabuilder.Schema{
+		"schema1": map[string]*schemabuilder.Schema{
+			"schema1": buildTestSchema1(),
+		},
+		"schema2": map[string]*schemabuilder.Schema{
+			"schema2": buildTestSchema2(),
+		},
+	}
+
+	builtSchemas := make(serviceSchemas)
+	for service, versions := range schemas {
+		builtSchemas[service] = make(map[string]*introspectionQueryResult)
+		for version, schema := range versions {
+			builtSchemas[service][version] = extractSchema(t, schema.MustBuild())
+		}
+	}
+	merged, err := convertVersionedSchemas(builtSchemas)
+	require.NoError(t, err)
+
+	f, err := newFlattener(merged.Schema)
+	return &Executor{
+		flattener: f,
+		schema: merged,
+	}, nil
+}
+
+func TestPlanner(t *testing.T) {
+	e,err := setupExecutor(t)
+	require.NoError(t, err)
+
+	testCases := []struct {
+		Name   string
+		Input  string
+		Output []*Plan
+	}{
+		{
+			Name: "query with fields from the same service",
+			Input: `query Test {
+				s1fff {
+					name
+				}
+			}`,
+			Output: []*Plan{
+				{
+					Service: "schema1",
+					Type:    "Query",
+					Kind:    "query",
+					SelectionSet: mustParse(`{
+						s1fff {
+							name
+						}
+					}`),
+					After: nil,
+				},
+			},
+		},
+		{
+			Name: "query with fields from two seperate services",
+			Input: `query Test {
+				s1fff {
+					name
+					s2ok
+				}
+			}`,
+			Output: []*Plan{
+				{
+					Service: "schema1",
+					Type:    "Query",
+					Kind:    "query",
+					SelectionSet: mustParse(`{
+						s1fff {
+							name
+							s2ok
+						}
+					}`),
+					After: []*Plan{
+						{
+							Path: []PathStep{
+								{Kind: KindField, Name: "s1fff"},
+							},
+							Type:    "Foo",
+							Kind:    "query",
+							Service: "schema2",
+							SelectionSet: mustParse(`{
+								s2ok
+							}`),
+						},
+					},
+				},
+			},
+		},
+		{
+			Name: "query with fields from nested levels between two services",
```

Comment:
this test looks identical to the one above. get rid of it? i think the one below covers it